rfd/0127-encrypted-session-recordings.md

```diff
@@ -0,0 +1,489 @@
+  // session recording
+  SessionRecordingConfigStatusV2 Status = 6 [
+    (gogoproto.jsontag) = "status",
+    (gogoproto.nullable) = true
+  ]
+}
+```
+
+```proto
+// api/proto/teleport/clusterconfig/v1/clusterconfig_service.proto
+
+// existing messages omitted
+
+message RotateSessionRecordingConfigRequest {}
+message RotateSessionRecordingConfigResponse {}
+
+// ClusterConfigService provides methods to manage cluster configuration resources.
+service ClusterConfigService {
+  // existing RPCs omitted
+
+  // RotateSessionRecordingConfigKeys rotates the keys associated with encrypting and
+  // decrypting session recordings.
+  rpc RotateSessionRecordingConfig(RotateSessionRecordingConfigRequest) returns (RotateSessionRecordingConfigResponse);
+}
+```
+
+```proto
+// api/proto/teleport/legacy/client/proto/authservice.proto
+
+// ClusterConfigService provides methods to manage cluster configuration resources.
+service ClusterConfigService {
+  // existing messages omitted
+
+  // EncryptedSessionRecordingChunk is an individual chunk of an encrypted 
+  // session recording .tar file.
+  message EncryptedSessionRecordingChunk {
+    // SessionID the recording relates to.
+    string SessionID = 1;
+    // ChunkIndex is the ordered index applied to the chunk.
+    int64 ChunkIndex = 2;
+    // Chunk is the encrypted chunk of session recording data being uploaded.
+    bytes Chunk = 3;
+  }
+
+  message UploadEncryptedSessionRecordingResponse {}
+
+  service AuthService {
+    // existing RPCs omitted
+
+    // UploadEncryptedSessionRecording is used to upload encrypted .tar files
+    // containing session recording events into long term storage.
+    rpc UploadEncryptedSessionRecording(stream EncryptedSessionRecordingChunk) returns (UploadEncryptedSessionRecordingResponse);
+    
+  }
+}
+```
+
+### Session Recording Modes
+
+There are four session recording modes that describe where the recordings are
+captured and how they're shipped to long term storage.
+- `proxy-sync`
+- `proxy`
+- `node-sync`
+- `node`
+Where the recordings are collected is largely unimportant to the encryption
+strategy, but whether or not they are handled async or sync has different
+requirements.
+
+In sync modes the session recording data is written immediately to the auth
+service without intermediate disk writes. The auth service then handles
+batching and multipart upload of events to long term storage. In order to
+ensure resilience to partial uploads of parts, each batch will need to be
+encrypted individually on the auth service. This will likely mean wrapping the
+`gzipWriter` [used by slices](https://github.com/gravitational/teleport/blob/master/lib/events/stream.go#L669)
+with an `age` encryption writer.
+
+In async modes the session recording data is written to intermediate `.part`
+files. These files are collected until they're ready for upload and are then
+combined into a single `.tar` file. In order to encrypt individual parts, we
```

Comment:
is this what already happens, all parts being into a single .tar for upload? if it's new i wonder if it could be a problem if the .tar is quite big, could take a lot of memory

Reply:
It is but the `.tar` is parsed into individual events that are sent to the auth service in more or less the same way `sync` modes work. Since the `.tar` is already encrypted that won't be possible, but we can still chunk the upload to the auth service.